autoload/vsnip/variable.vim

```diff
@@ -0,0 +1,171 @@
+let s:variables = {}
+
+"
+" vsnip#variable#register
+"
+function! vsnip#variable#register(name, func, ...) abort
+  let l:option = get(a:000, 0, {})
+  let s:variables[a:name] = {
+  \   'func': a:func,
+  \   'once': get(l:option, 'once', v:false)
+  \ }
+endfunction
+
+"
+" vsnip#variable#get
+"
+function! vsnip#variable#get(name) abort
+  return get(s:variables, a:name, v:null)
+endfunction
+
+"
+" Register built-in variables.
+"
+" @see https://code.visualstudio.com/docs/editor/userdefinedsnippets#_variables
+"
+
+function! s:TM_SELECTED_TEXT(context) abort
+  let l:selected_text = vsnip#selected_text()
+  if !empty(l:selected_text)
+    " TODO: fix indentation with a:context
+    return l:selected_text
```

Comment:
Currently, `s:Variable.resolve` can't provide detailed context to fix the indentation so we should re-design `s:Variable.resolve` timing.